Handle Errors when following "next" link when fetching collection

anthonyfalzetti opened the pull request:
While following the next link it times out. Rather than return a partial collection it should raise an error. 

This PR will: 
* Create a new Error type NoSuchLinkError
* When an unsuccessful response is returned raise CollectionError

Example Stacktrace: 
```
** (ExHal.CollectionError) Failed to fetch next link due to timeout
    lib/exhal/collection.ex:24: anonymous fn/1 in ExHal.Collection.to_stream/1
    (elixir) lib/stream.ex:1377: Stream.do_resource/5
    (elixir) lib/stream.ex:1553: Enumerable.Stream.do_each/4
    (elixir) lib/enum.ex:2979: Enum.map/2
```

Commit:
Handle Errors in Collections

* Create a new Error type NoSuchLinkError
* When an unsuccessful response is returned raise CollectionError

heyZeus commented on lib/exhal/collection.ex:
```diff
@@ -14,7 +14,17 @@ defmodule ExHal.Collection do
+          {:error, %ExHal.Error{reason: msg}} ->
+            raise ExHal.CollectionError, message: "Failed to fetch next link due to #{msg}"
+          {:error, %ExHal.Error{reason: msg}, _headers} ->
+            raise ExHal.CollectionError, message: "Failed to fetch next link due to #{msg}"
```
Now `to_stream` raises an exception. Maybe a new function called `to_stream!`

pezra replied:
I think it makes sense for a stream to raise if there is a problem traversing the list.

anthonyfalzetti replied:
👍

pezra commented on lib/exhal/navigation.ex:
```diff
@@ -25,13 +25,13 @@ defmodule ExHal.Navigation do
   Follows all links of a particular rel in a HAL document.
 
-  Returns `[{:ok, %ExHal.Document{...}, %ExHal.ResponseHeader{...}}, {:error, %ExHal.Error{...}, ...]`
+  Returns `[{:ok, %ExHal.Document{...}, %ExHal.ResponseHeader{...}}, {:error, %ExHal.NoSuchLinkError{...}, {:error, %ExHal.Error{...} ...]`
```
I wonder if `NoSuchLinkError` should be embedded in a `List`. It is not the same as failing to fetch one target of a compound link. If the document doesn't have a link that's it.

I see a couple of reasonable choices when the link doesn't exist:
 1. return an empty list
 2. return an error tuple like `{:error, %ExHal.NoSuchLinkError{...}}`

If we are going to make a breaking change to the API (which this PR does) we should do one of those things.

anthonyfalzetti replied:
I opted to go with `{:error, %ExHal.NoSuchLinkError{...}}` and added test coverage

pezra reviewed (changes requested):
:+1: to raising a error if fetching the next page in a collection fails. However, i don't think this change accomplished the intended effect. I'd like to see some more tests around the target edge case.

Commit:
Add test to CollectionTest

Evaluate the client at runtime rather than at compile time similar to how it is used for Form.
This was needed for Mocking a timeout. Alternative would be to build out a config/test.exs.

Commit:
Switch client_module to be cyphered at runtime. This is how it is handled in form.ex

Commit:
When link is not present on doc return {:error, %ExHal.NoSuchLinkError} as default

anthonyfalzetti commented on test/test_helper.exs:
```diff
@@ -3,5 +3,3 @@ Application.ensure_all_started(:mox)
 
 Mox.defmock(ExHal.ClientMock, for: ExHal.Client)
-
-Application.put_env(:exhal, :client, ExHal.ClientMock)
```
Removed from test_helper.exs since not all tests wish to mock using Mox. Moving Navigation to use runtime client_module this causes issues in testing.

anthonyfalzetti commented on test/exhal/form_test.exs:
```diff
@@ -9,6 +9,13 @@ defmodule ExHal.FormTest do
+  setup do
+    Application.put_env(:exhal, :client, ExHal.ClientMock)
+    on_exit fn ->
+      Application.put_env(:exhal, :client, ExHal.Client)
```
When running `mix test test/` if the client is not returned to `ExHal.Client` it fails tests downstream.

anthonyfalzetti commented on test/exhal/collection_test.exs:
```diff
@@ -84,6 +86,31 @@ defmodule ExHal.CollectionTest do
     assert ExHal.to_stream(doc) |> is_a_stream
   end
 
+  describe "Non Hal Responses" do
```
Had to wrap this in a describe block so the setup doesn't affect other tests. Open to a different way to do this.

Commit:
Test async: false

In CI the test appears to fail due to incorrect tests attempting to get
ClientMock. Setting to async: false to see if it goes away.

adherr commented on lib/exhal/navigation.ex:
```diff
@@ -1,8 +1,8 @@
 defmodule ExHal.Navigation do
-  @client_module Application.get_env(:exhal, :client, ExHal.Client)
+  # client_module() Application.get_env(:exhal, :client, ExHal.Client)
```
Probably want to take this out if it's unused

anthonyfalzetti replied:
correct. Consider it gone.

alakra commented on lib/exhal/collection.ex:
```diff
@@ -14,7 +14,17 @@ defmodule ExHal.Collection do
       fn follow_result ->
         case follow_result do
-          {:error, _} -> {:halt, follow_result}
+          # End of Pagination
```
Since this is now raising, we should probably document that behavior. Here's a proposal for an updated `@doc`:

```
@doc """
Returns a stream that iterates over the collection represented by `a_doc`.

In the case of failures during processing the stream, the stream will halt if 
an `ExHal.NoSuchLinkError` error occurs when following the next resource.  

If an `ExHal.Error` occurs when requesting the next resource, an 
`ExHal.CollectionError` will be raised to the caller.
"""
```

anthonyfalzetti replied:
Thanks for the suggestion. Updated.

adherr replied:
I think that second paragraph is a little misleading because the `NoSuchLinkError` is not a failure, it's our signal to stop iterating. Maybe add this to the first paragraph?

```
Iteration halts when there is no further `next` link to follow.
```

alakra replied:
That's a good point. I second that change.

Commit:
Update comments on Collection.stream/1

pezra commented on lib/exhal/collection.ex:
```diff
@@ -7,14 +7,28 @@ defmodule ExHal.Collection do
+  Iteration halts when there is no further `next` link to follow.
+
+  If an `ExHal.Error` occurs when requesting the next resource, an
+  `ExHal.CollectionError` will be raised to the caller.
```
Maybe be a little clearer that a collection error will be raised only when following the next link to the the next page? This currently reads a little like it might raise a collection error if one of the members of the collection is unfetchable.

comverge-pair replied:
updated

anthonyfalzetti replied:
updated

anthonyfalzetti commented:
Going to merge and release after work today.
